File: editor-server/src/resolvers/color.ts
import {
  Resolver,
  Query,
  Arg,
  Ctx,
  Mutation,
  Publisher,
  PubSub,
  Int,
} from "type-graphql";

import { Topic } from "./subscriptions/topic";
import { ColorPayload, colorMutation } from "./subscriptions/color";
import { ColorMap } from "./types/colorMap";
import { ColorResponse } from "./response/colorResponse";
import { IControl, IControlFrame, TContext } from "../types/global";
import {
  Color,
  ColorCreateInput,
  ColorUpdateInput,
} from "../../prisma/generated/type-graphql";

@Resolver()
class ColorResolver {
  @Query(() => ColorMap)
  async colorMap(@Ctx() ctx: TContext) {
    const colors = await ctx.prisma.color.findMany();
    return { colorMap: colors };
  }

  @Mutation(() => Color)
  async editColor(
    @PubSub(Topic.Color) publish: Publisher<ColorPayload>,
    @Arg("id", () => Int) id: number,
    @Arg("data") data: ColorUpdateInput,
    @Ctx() ctx: TContext
  ) {
    // Check colorCode dimension
    const { colorCode } = data;
    if (colorCode) {
      const { set } = colorCode;
      if (set && set.length !== 3) {
        throw new Error(`colorCode dimension should be 3, get ${set.length}`);
      }
    }

    await ctx.prisma.color.findUniqueOrThrow({
      where: { id },
    });

    const colorData = await ctx.prisma.color.update({
      where: { id },
      data: data,
    });

    const payload: ColorPayload = {
      mutation: colorMutation.UPDATED,
      id: id,
      color: colorData.color,
      colorCode: colorData.colorCode,
      editBy: ctx.userId,
    };
    await publish(payload);

    return colorData;
  }

  @Mutation(() => Color)
  async addColor(
    @PubSub(Topic.Color) publish: Publisher<ColorPayload>,
    @Arg("color") colorInput: ColorCreateInput,
    @Ctx() ctx: TContext
  ) {
    // Check colorCode dimension
    const { colorCode } = colorInput;
    if (colorCode) {
      const { set } = colorCode;
      if (set.length !== 3) {
        throw new Error(`colorCode dimension should be 3, get ${set.length}`);
      }
    }

    // Create new Color, if "color" duplicate => create() will throw error
    const color = await ctx.prisma.color.create({
      data: colorInput,
    });

    // publish
    const payload: ColorPayload = {
      mutation: colorMutation.CREATED,
      id: color.id,
      color: color.color,
      colorCode: color.colorCode,
      editBy: ctx.userId,
    };
    await publish(payload);

    return color;
  }

  // @Mutation(() => String)
  // async renameColor(
  //   @PubSub(Topic.Color) publish: Publisher<ColorPayload>,
  //   @Arg("originalColor") originalColor: string,
  //   @Arg("newColor") newColor: string,
  //   @Ctx() ctx: TContext
  // ) {
  //   // check if new color name exists, if true => throw error
  //   const existedNewColor = await ctx.prisma.color.findUnique({
  //     where: { color: newColor },
  //   });
  //   if (existedNewColor) {
  //     throw new Error(
  //       `color name: ${newColor}/code: ${existedNewColor.colorCode} existed`
  //     );
  //   }

  //   // check if old color name exists, if not => throw error
  //   await ctx.prisma.color.findUniqueOrThrow({
  //     where: { color: originalColor },
  //   });

  //   // check if color is used in ControlData
  //   const checkColorInControl = await ctx.prisma.controlData.findMany({
  //     where: { value: { path: ["color"], equals: originalColor } },
  //   });
  //   if (checkColorInControl.length !== 0) {
  //     throw new Error(`color is used in ControlData`);
  //   }

  //   await ctx.prisma.color.update({
  //     where: { color: originalColor },
  //     data: { color: newColor },
  //   });

  //   // publish
  //   const payload: ColorPayload = {
  //     mutation: colorMutation.RENAMED,
  //     color: originalColor,
  //     renameColor: newColor,
  //     editBy: ctx.userId,
  //   };
  //   await publish(payload);

  //   return newColor;
  // }

  @Mutation(() => ColorResponse)
  async deleteColor(
    @PubSub(Topic.Color) publish: Publisher<ColorPayload>,
    @Arg("id", () => Int) id: number,
    @Ctx() ctx: TContext
  ) {
    try {
      // check if color id exists
      const existedColor = await ctx.prisma.color.findUniqueOrThrow({
        where: { id: id },
      });

      // check whether color is using in Control
      const checkControl = await ctx.prisma.controlData.findMany({
        where: { value: { path: ["color"], equals: id } },
      });
      if (checkControl.length !== 0) {
        let checkControlFrames: number[] = checkControl.map(
          (control) => control.frameId
        );
        checkControlFrames = checkControlFrames.sort(function (a, b) {
          return a - b;
        });
        let frame = 0;
        const ids: number[] = [];
        checkControlFrames.map((controlFrame) => {
          if (controlFrame !== frame) {
            ids.push(controlFrame);
            frame = controlFrame;
          }
        });
        return {
          id: id,
          ok: false,
          msg: `color is used in frames ${ids}`,
        };
      }

      // TODO: check whether color is using in LEDEffect
      const LEDEffects = await ctx.prisma.lEDEffect.findMany({
        select: { frames: true, id: true },
      });
      const ids: number[] = [];
      LEDEffects.map((LEDEffect) => {
        const { frames } = LEDEffect;
        const ledId = LEDEffect.id;
        frames.map((frame: any) => {
          const LEDs = frame.LEDs;
          LEDs.map((led: any) => {
            if (led[0].toString() === id.toString() && !ids.includes(ledId)) {
              ids.push(ledId);
            }
          });
        });
      });
      if (ids.length !== 0)
        return {
          id: id,
          ok: false,
          msg: `color is used in led ${ids}`,
        };

      // Delete Color
      const deleteColor = await ctx.prisma.color.delete({
        where: { id: id },
      });
      const payload: ColorPayload = {
        mutation: colorMutation.DELETED,
        id: id,
        editBy: ctx.userId,
      };
      await publish(payload);
      return { id: id, ok: true };
    } catch (error) {
      console.log(error);
      return Object.assign(
        { id: id },
        { ok: false, msg: `color id ${id} doesn't existed` }
      );
    }
  }
}

export default ColorResolver;

File: editor-server/src/graphql/mutations/color.rs
//! Color mutation methods.
use crate::db::types::color::ColorData;
use crate::graphql::types::led::{LEDEffectData, LEDEffectFrame};
use crate::graphql::{
    subscriptions::color::{ColorMutationMode, ColorPayload},
    subscriptions::led::LEDPayload,
    subscriptor::Subscriptor,
    types::color::Color,
};
use crate::types::global::UserContext;
use crate::utils::revision::update_revision;

use async_graphql::{Context, InputObject, Object, Result as GQLResult, SimpleObject};
use itertools::Itertools;

// TODO: Remove this after all done
#[derive(InputObject, Default, Debug)]
pub struct StringFieldUpdateOperationsInput {
    pub set: String,
}

#[derive(InputObject, Default, Debug)]
pub struct ColorUpdateColorCodeInput {
    pub set: Vec<i32>,
}

#[derive(InputObject, Default)]
pub struct ColorUpdateInput {
    pub color: StringFieldUpdateOperationsInput,
    pub color_code: ColorUpdateColorCodeInput,
}

#[derive(InputObject, Default, Debug)]
pub struct ColorCreateColorCodeInput {
    pub set: Vec<i32>,
}

#[derive(InputObject, Default)]
pub struct ColorCreateInput {
    pub color: String,
    pub color_code: ColorCreateColorCodeInput,
    pub auto_create_effect: Option<bool>,
}

#[derive(SimpleObject, Default)]
pub struct ColorResponse {
    pub id: i32,
    pub msg: String,
    pub ok: bool,
}

#[derive(Default)]
pub struct ColorMutation;

#[Object]
impl ColorMutation {
    async fn edit_color(
        &self,
        ctx: &Context<'_>,
        id: i32,
        data: ColorUpdateInput,
    ) -> GQLResult<Color> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();

        tracing::info!("Mutation: editColor");

        let led_effect = sqlx::query!(
            r#"
                SELECT name FROM Color
                WHERE id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await;

        sqlx::query!(
            r#"
                UPDATE Color SET name = ?, r = ?, g = ?, b = ?
                WHERE id = ?;
            "#,
            &data.color.set,
            data.color_code.set[0],
            data.color_code.set[1],
            data.color_code.set[2],
            id
        )
        .execute(mysql)
        .await?;

        if let Ok(led_effect) = led_effect {
            sqlx::query!(
                r#"
                    UPDATE LEDEffect SET name = ?
                    WHERE name = ?;
                "#,
                &data.color.set,
                led_effect.name
            )
            .execute(mysql)
            .await?;
        }

        let color_payload = ColorPayload {
            mutation: ColorMutationMode::Updated,
            id,
            color: Some(data.color.set.clone()),
            color_code: Some(data.color_code.set.clone()),
            edit_by: context.user_id,
            // edit_by: 0,
        };

        Subscriptor::publish(color_payload);

        update_revision(mysql).await?;

        let color = Color {
            id,
            color: data.color.set,
            color_code: data.color_code.set,
        };

        Ok(color)
    }

    async fn add_color(&self, ctx: &Context<'_>, color: ColorCreateInput) -> GQLResult<Color> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();

        let id = sqlx::query!(
            r#"
                INSERT INTO Color (name, r, g, b)
                VALUES (?, ?, ?, ?);
            "#,
            &color.color,
            color.color_code.set[0],
            color.color_code.set[1],
            color.color_code.set[2]
        )
        .execute(mysql)
        .await?
        .last_insert_id() as i32;

        let color_payload = ColorPayload {
            mutation: ColorMutationMode::Created,
            id,
            color: Some(color.color.clone()),
            color_code: Some(color.color_code.set.clone()),
            edit_by: context.user_id,
        };

        Subscriptor::publish(color_payload);

        if color.auto_create_effect.unwrap_or(false) {
            let model_parts = sqlx::query!(
                r#"
                    SELECT
                        Model.id,
                        Model.name AS model_name,
                        Part.id AS part_id,
                        Part.name AS part_name,
                        Part.length AS part_length
                    FROM Model
                    INNER JOIN Part ON Model.id = Part.model_id
                    WHERE Part.type = 'LED';
                "#
            )
            .fetch_all(mysql)
            .await?
            .into_iter()
            .map(|row| {
                let length = row.part_length.unwrap_or(0);
                (row.id, row.part_id, length, row.model_name, row.part_name)
            })
            .collect_vec();

            let create_effect_futures = model_parts.iter().map(|model_part| {
                sqlx::query!(
                    r#"
                        INSERT INTO LEDEffect (name, model_id, part_id)
                        VALUES (?, ?, ?);
                    "#,
                    color.color.clone(),
                    model_part.0,
                    model_part.1
                )
                .execute(mysql)
            });

            let create_effect_results = futures::future::join_all(create_effect_futures)
                .await
                .into_iter()
                .filter_map(|result| result.ok())
                .collect_vec();

            let mut create_states_futures = Vec::new();
            create_effect_results
                .iter()
                .zip(&model_parts)
                .for_each(|(result, model_part)| {
                        for pos in 0..model_part.2 {
                            create_states_futures.push(
                                sqlx::query!(
                                    r#"
                                        INSERT INTO LEDEffectState (effect_id, position, color_id, alpha)
                                        VALUES (?, ?, ?, ?)
                                        ON DUPLICATE KEY UPDATE color_id = VALUES(color_id), alpha = VALUES(alpha);
                                    "#,
                                    result.last_insert_id() as i32,
                                    pos,
                                    id,
                                    255
                                )
                                .execute(mysql),
                            )
                        }
                });

            futures::future::join_all(create_states_futures).await;

            let create_effects = create_effect_results
                .iter()
                .zip(model_parts)
                .map(|(result, model_part)| {
                    let effect_id = result.last_insert_id() as i32;
                    let frames = (0..model_part.2)
                        .map(|pos| LEDEffectFrame {
                            leds: (0..model_part.2).map(|_| [id, 255]).collect_vec(),
                            fade: false,
                            start: pos,
                        })
                        .collect_vec();

                    LEDEffectData {
                        id: effect_id,
                        name: color.color.clone(),
                        model_name: model_part.3,
                        part_name: model_part.4,
                        repeat: 0,
                        frames,
                    }
                })
                .collect_vec();

            let led_payload = LEDPayload {
                create_effects,
                update_effects: Vec::new(),
                delete_effects: Vec::new(),
            };

            Subscriptor::publish(led_payload);
        }

        update_revision(mysql).await?;

        let color = Color {
            id,
            color: color.color,
            color_code: color.color_code.set,
        };

        Ok(color)
    }

    #[allow(unused)]
    async fn delete_color(&self, ctx: &Context<'_>, id: i32) -> GQLResult<ColorResponse> {
        let context = ctx.data::<UserContext>()?;
        let app_state = &context.clients;

        let mysql = app_state.mysql_pool();

        let color = sqlx::query_as!(
            ColorData,
            r#"
                SELECT * FROM Color
                WHERE id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await?;

        let check_color = sqlx::query!(
            r#"
                SELECT COUNT(ControlData.frame_id) AS count
                FROM ControlData
                WHERE ControlData.color_id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await?
        .count;

        let check_effect = sqlx::query!(
            r#"
                SELECT COUNT(ControlData.frame_id) AS count
                FROM LEDEffectState
                INNER JOIN ControlData 
                    ON ControlData.effect_id = LEDEffectState.effect_id
                WHERE LEDEffectState.color_id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await?
        .count;

        if check_color + check_effect > 0 {
            return Ok(ColorResponse {
                id: 0,
                msg: "Color is used in control data with id.".to_string(),
                ok: false,
            });
        }

        let _ = sqlx::query!(
            r#"
                DELETE FROM Color
                WHERE id = ?;
            "#,
            id
        )
        .execute(mysql)
        .await?;

        let color_payload = ColorPayload {
            mutation: ColorMutationMode::Deleted,
            id,
            color: None,
            color_code: None,
            edit_by: context.user_id,
        };

        Subscriptor::publish(color_payload);

        let delete_effects = sqlx::query!(
            r#"
                SELECT
                    DISTINCT LEDEffect.id,
                    Model.name AS model_name,
                    Part.name AS part_name
                FROM LEDEffectState
                INNER JOIN LEDEffect
                    ON LEDEffectState.effect_id = LEDEffect.id
                INNER JOIN Model ON LEDEffect.model_id = Model.id
                INNER JOIN Part ON LEDEffect.part_id = Part.id
                WHERE LEDEffectState.color_id = ?;
            "#,
            id
        )
        .fetch_all(mysql)
        .await?;

        let delete_effect_futures = delete_effects.iter().map(|effect| {
            sqlx::query!(
                r#"
                    DELETE FROM LEDEffect
                    WHERE id = ?;
                "#,
                effect.id
            )
            .execute(mysql)
        });

        futures::future::join_all(delete_effect_futures).await;

        let delete_effects = delete_effects
            .iter()
            .map(|effect| LEDEffectData {
                id: effect.id,
                name: color.name.clone(),
                model_name: effect.model_name.clone(),
                part_name: effect.part_name.clone(),
                repeat: 0,
                frames: Vec::new(),
            })
            .collect_vec();

        let led_payload = LEDPayload {
            create_effects: Vec::new(),
            update_effects: Vec::new(),
            delete_effects,
        };

        Subscriptor::publish(led_payload);

        update_revision(mysql).await?;

        Ok(ColorResponse {
            id,
            msg: "Color deleted.".to_string(),
            ok: true,
        })
    }
}
